src/osd/sdl/osdsdl.h

```diff
@@ -32,6 +32,13 @@
 #define SDLOPTION_LIGHTGUNINDEX         "lightgun_index"
 #endif
 
+#if (USE_EVDEV)
+#define SDLOPTION_KEYBOARD_DEV          "keyboard_dev"
+#define SDLOPTION_JOYSTICK_DEV          "joystick_dev"
+#define SDLOPTION_MOUSE_DEV             "mouse_dev"
+#define SDLOPTION_LIGHTGUN_DEV          "lightgun_dev"
+#endif
+
```

Comment:
This isn’t the direction we’re taking for things like this – please ensure you’re exposing predictable IDs for devices so they can be remapped using controller configuration files.

Reply:
Gotcha.  First attempt at messing with the mame code so I'm not all that aware of "this is what we're aiming for today" as opposed to modeling on "this is how we did things yesteryear".  :-)

I'll axe the mapping options entirely and concentrate on stable unique IDs instead.

Reply:
There remains an issue with lightguns, however, and no obvious way to fix it without introducing a command-line option that I can think of: they present themselves to the OS as mice.  Short of a way to override that, they'll show up as such.

On the other hand, evdev doesn't _care_ about what is a mouse or whatever.  As far as it's concerned everything is just a device with an arbitrary number of buttons and axes and it'll happily allow you to use it however you want -- a "keyboard" is just a joystick with no axes and lots of buttons.  :-)

The only reason I care at all in the code is because _Mame_ makes a distinction in the UI, so I guess from the available events whether it should show up as a "Keyboard", "Mouse" or "Joystick" -- but I can't guess "Lightgun".

Reply:
The reasons would be better automatic mappings on analog axes (we really need to work on that).  Wheels and flight controls are not really compatible, usage-wise :-)

Reply:
Lightgun and mouse inputs are logically different though:
* A lightgun (or a lightpen, or a digitiser tablet for that matter) gives you pointer acquired (gun pointed at screen, or pen in proximity), absolute X/Y locations (referenced to the screen), and some additional action controls (gun triggers, pen pressure, pen angle).
* A mouse or trackball gives relative X/Y movement, and additional button state.

Now in practice, PC lightguns often produce pointer movement, and digitiser tablets often provide relative movement as a fallback in case the OS doesn’t have suitable drivers.  As a result, in MAME, a lightgun peripheral may end up appearing as a mouse *and* a lightgun.

The separation between device classes is useful for a few reasons. As well as automatic input assignments that @galibert mentioned, you can specify different modules to handle different classes of device.  For example, suppose you’re using Windows on a notebook computer with a precision touchpad.  This won’t produce RawInput mouse events, so you may want to switch your mouse provider to the one that uses Win32 window messages, while still using the default RawInput keyboard provider.